File: sample-app/iosApp/iosApp/AuthViewModel.swift
//
//  AuthViewModel.swift
//  iosApp
//
//

import Foundation
import Auth0

@MainActor
@Observable
final class AuthViewModel {

    enum State: Equatable {
        // Initial state while the app checks storage for a saved session; the
        // splash overlay is shown until this resolves to success or idle.
        case restoring
        case idle
        case loading
        case success(Credentials)
        case failure(any Auth0Error)

        static func == (lhs: State, rhs: State) -> Bool {
            switch (lhs, rhs) {
                  case (.restoring, .restoring), (.idle, .idle), (.loading, .loading):
                      return true
                  case let (.success(l), .success(r)):
                      return l.isEqual(r)
                  case let (.failure(l), .failure(r)):
                return (l as? NSObjectProtocol)?.isEqual(r) ?? false
                  default:
                      return false
                  }
        }
    }

    enum LoginMethod {
        case embedded
        case webAuth
        case passkey
    }

    // Sign-up outcome, kept separate from `State` on purpose: createUser mints no
    // tokens, so a successful sign-up must NOT trip the global success→Welcome
    // navigation. This drives its own screen (the confirmation) instead.
    enum SignupState: Equatable {
        case idle
        case loading
        case success(DatabaseUser)
        case failure(any Auth0Error)

        static func == (lhs: SignupState, rhs: SignupState) -> Bool {
            switch (lhs, rhs) {
            case (.idle, .idle), (.loading, .loading):
                return true
            case let (.success(l), .success(r)):
                return l == r
            case let (.failure(l), .failure(r)):
                return (l as? NSObjectProtocol)?.isEqual(r) ?? false
            default:
                return false
            }
        }
    }

    private(set) var state: State = .restoring

    private(set) var signupState: SignupState = .idle

    // Remembers the credentials entered on the sign-up screen so the confirmation
    // screen's "Log in" button can complete the sign-up → login hop without asking
    // the user to retype them. Cleared by resetSignup().
    private var lastSignup: (email: String, password: String, connection: String)?

    // Tracks how the current session was established, so logout only performs the
    // browser round-trip for Web Auth sessions (embedded login holds no SSO cookie).
    private var loginMethod: LoginMethod?

    let isConfigured: Bool

    private let client: (any AuthenticationClient)?

    private let webClient: (any WebAuthClient)?

    private let credentialsManager: (any CredentialsManager)?

    private let audience = "https://firstresourceserver/"

    init(domain: String, clientId: String) {
        isConfigured = !domain.isEmpty && !clientId.isEmpty
        guard isConfigured else {
            client = nil
            webClient = nil
            credentialsManager = nil
            // No SDK client to check storage with, so there is nothing to restore.
            state = .idle
            return
        }
        // Kotlin default arguments do NOT cross into Swift, so every parameter
        // the Kotlin side defaults must be supplied explicitly here. These match
        // the SDK's Kotlin defaults (NetworkingConfiguration()).
        let configuration = NetworkingConfiguration(
            logLevel: .body,
            connectTimeoutMillis: 10_000,
            requestTimeoutMillis: 10_000,
            defaultHeaders: [:]
        )
        let account = Auth0Account(
            clientId: clientId,
            domain: domain,
            configuration: configuration,
            useDPoP: false
        )
        let userAgent = Auth0UserAgent.companion.default()
        client = authenticationClient(account: account, userAgent: userAgent)
        webClient = webAuthClient(account: account, userAgent: userAgent)
        // Uses the account-scoped store key and platform (Keychain) storage; the
        // single-argument overload exists so Swift needs no internal factories.
        // Module-qualified because the stored property above shadows the global
        // factory function of the same name inside this class.
        credentialsManager = Auth0.credentialsManager(account: account)
    }

    // Checks the Keychain for a saved session on launch. A restored session goes
    // straight to the logged-in screen; anything else (no credentials, or an
    // expired token with no refresh token) is not an error to show — it just
    // means the user needs to sign in, so we fall back to idle (the chooser).
    func restoreSession() async {
        guard let credentialsManager else {
            state = .idle
            return
        }
        do {
            let result = try await credentialsManager.getCredentials(
                scope: nil,
                minTtl: 30,
                parameters: [:],
                headers: [:],
                forceRefresh: false
            )
            switch onEnum(of: result) {
            case .success(let success):
                if let credentials = success.data as? Credentials {
                    state = .success(credentials)
                } else {
                    state = .idle
                }
            case .failure:
                state = .idle
            }
        } catch {
            state = .idle
        }
    }

    func login(
        email: String,
        password: String,
        realm: String,
        options: RequestOptions = RequestOptions(
            parameters: [:],
            headers: [:],
            retryPolicy: RetryPolicy.companion.None
        )
    ) async {
        guard let client else { return }
        state = .loading
        do {
            // Request offline_access so the tenant issues a refresh token, letting
            // getCredentials() renew an expired access token on a later launch.
            let result = try await client.login(
                usernameOrEmail: email,
                password: password,
                realm: realm,
                audience: audience,
                scope: "openid profile email offline_access",
                options: options
            )
            // SKIE turns the Kotlin sealed `Result` into a Swift enum we can
            // switch over exhaustively. The payload generic erases to AnyObject,
            // so we downcast to the concrete SDK types.
            switch onEnum(of: result) {
            case .success(let success):
                if let credentials = success.data as? Credentials {
                    loginMethod = .embedded
                    await saveCredentials(credentials)
                    state = .success(credentials)
                } else {
                    state = .idle
                }
            case .failure(let failure):
                if let error = failure.error as? AuthenticationError {
                    state = .failure(error)
                } else {
                    state = .idle
                }
            }
        } catch {
            // The suspend bridge is `async throws`; domain failures come back as
            // Result.Failure (handled above), so a thrown error here is unexpected
            // (e.g. cancellation). Reset for this sample.
            state = .idle
        }
    }
    
    
    func webLogin() async {
        guard let webClient else { return }
        state = .loading
        do {
                     let result = try await webClient.login(
                         options: LoginOptions(
                             scope: "openid profile email offline_access",
                             audience: audience,
                             connection: nil,
                             organization: nil,
                             prompt: nil,
                             maxAge: nil,
                             redirectUri: nil,
                             scheme: nil,
                             ephemeral: false,
                             extraParameters: [:]
                         )
                     )
                     switch onEnum(of: result) {
                     case .success(let success):
                         if let credentials = success.data as? Credentials {
                             loginMethod = .webAuth
                             await saveCredentials(credentials)
                             state = .success(credentials)
                         } else {
                             state = .idle
                         }
                     case .failure(let failure):
                         if let error = failure.error as? WebAuthError {
                             state = .failure(error)
                         } else {
                             state = .idle
                         }
                     }
                 } catch {
                     state = .idle
                 }
    }

    // Creates a database user. On success we land on the confirmation screen
    // (via signupState); createUser returns a DatabaseUser, not tokens, so the
    // login state is untouched and the user explicitly opts into logging in next.
    func createUser(email: String, password: String, connection: String) async {
        guard let client else { return }
        signupState = .loading
        do {
            let result = try await client.createUser(
                profile: SignupProfile(
                    email: email,
                    phoneNumber: nil,
                    username: nil,
                    name: nil,
                    givenName: nil,
                    familyName: nil,
                    nickname: nil,
                    picture: nil
                ),
                password: password,
                connection: connection,
                userMetadata: [:],
                options: defaultOptions()
            )
            switch onEnum(of: result) {
            case .success(let success):
                if let user = success.data as? DatabaseUser {
                    lastSignup = (email, password, connection)
                    signupState = .success(user)
                } else {
                    signupState = .idle
                }
            case .failure(let failure):
                if let error = failure.error as? AuthenticationError {
                    signupState = .failure(error)
                } else {
                    signupState = .idle
                }
            }
        } catch {
            signupState = .idle
        }
    }

    // Completes the sign-up → login hop using the credentials captured by
    // createUser. Delegates to login(), so a success flows through the normal
    // success→Welcome navigation.
    func completeSignupLogin() async {
        guard let signup = lastSignup else { return }
        await login(email: signup.email, password: signup.password, realm: signup.connection)
    }

    // Resets the sign-up screen before it is shown, so a stale confirmation from a
    // previous attempt never appears.
    func resetSignup() {
        signupState = .idle
        lastSignup = nil
    }

    // Passkey sign-up. The SDK is HTTP-only for passkeys, so the ceremony (the
    // Face ID sheet + new-passkey creation) is injected as a closure the caller
    // builds from the platform. Flow: signup challenge → run ceremony → send the
    // registration credential + the challenge's auth_session straight to
    // loginWithPasskey (no separate login challenge — verified against Auth0.Android).
    func passkeySignup(
        email: String,
        connection: String,
        runCeremony: (AuthnParamsPublicKey) async throws -> PublicKeyCredentials
    ) async {
        guard let client else { return }
        state = .loading
        do {
            let challengeResult = try await client.passkeySignupChallenge(
                profile: SignupProfile(
                    email: email,
                    phoneNumber: nil,
                    username: nil,
                    name: nil,
                    givenName: nil,
                    familyName: nil,
                    nickname: nil,
                    picture: nil
                ),
                userMetadata: [:],
                realm: connection,
                organization: nil,
                options: defaultOptions()
            )
            let challenge: PasskeyRegistrationChallenge
            switch onEnum(of: challengeResult) {
            case .success(let success):
                guard let value = success.data as? PasskeyRegistrationChallenge else {
                    state = .idle
                    return
                }
                challenge = value
            case .failure(let failure):
                state = (failure.error as? AuthenticationError).map(State.failure) ?? .idle
                return
            }

            let credential: PublicKeyCredentials
            do {
                credential = try await runCeremony(challenge.authParamsPublicKey)
            } catch let error as Auth0Error {
                state = .failure(error)
                return
            } catch {
                state = .idle
                return
            }

            await finishPasskeyLogin(
                authSession: challenge.authSession,
                credential: credential,
                realm: connection
            )
        } catch {
            state = .idle
        }
    }

    // Passkey login. Same shape as sign-up but with the login challenge and an
    // existing passkey; the assertion + auth_session go to loginWithPasskey.
    func passkeyLogin(
        connection: String,
        runCeremony: (AuthParamsPublicKey) async throws -> PublicKeyCredentials
    ) async {
        guard let client else { return }
        state = .loading
        do {
            let challengeResult = try await client.passkeyLoginChallenge(
                realm: connection,
                organization: nil,
                options: defaultOptions()
            )
            let challenge: PasskeyLoginChallenge
            switch onEnum(of: challengeResult) {
            case .success(let success):
                guard let value = success.data as? PasskeyLoginChallenge else {
                    state = .idle
                    return
                }
                challenge = value
            case .failure(let failure):
                state = (failure.error as? AuthenticationError).map(State.failure) ?? .idle
                return
            }

            let credential: PublicKeyCredentials
            do {
                credential = try await runCeremony(challenge.authParamsPublicKey)
            } catch let error as Auth0Error {
                state = .failure(error)
                return
            } catch {
                state = .idle
                return
            }

            await finishPasskeyLogin(
                authSession: challenge.authSession,
                credential: credential,
                realm: connection
            )
        } catch {
            state = .idle
        }
    }

    // Shared tail of both passkey flows: exchange the platform credential for
    // tokens. A passkey session is a local session (no browser cookie), so logout
    // treats it like embedded — hence loginMethod = .passkey.
    private func finishPasskeyLogin(
        authSession: String,
        credential: PublicKeyCredentials,
        realm: String
    ) async {
        guard let client else { return }
        do {
            let result = try await client.loginWithPasskey(
                authSession: authSession,
                authResponse: credential,
                realm: realm,
                organization: nil,
                audience: nil,
                scope: "openid email offline_access",
                options: defaultOptions()
            )
            switch onEnum(of: result) {
            case .success(let success):
                if let credentials = success.data as? Credentials {
                    loginMethod = .passkey
                    await saveCredentials(credentials)
                    state = .success(credentials)
                } else {
                    state = .idle
                }
            case .failure(let failure):
                state = (failure.error as? AuthenticationError).map(State.failure) ?? .idle
            }
        } catch {
            state = .idle
        }
    }

    // The SDK's RequestOptions default cannot cross from Kotlin into Swift, so
    // this rebuilds it explicitly — matching the login() parameter above.
    private func defaultOptions() -> RequestOptions {
        RequestOptions(
            parameters: [:],
            headers: [:],
            retryPolicy: RetryPolicy.companion.None
        )
    }

    func logout() {
        if loginMethod == .webAuth, let webClient {
            // Keep the current .success state (credentials stay on screen) during
            // the browser round-trip; only clear the view once logout succeeds.
            Task {
                do {
                    let result = try await webClient.logout(
                        options: LogoutOptions(
                            returnTo: nil,
                            scheme: nil,
                            federated: false,
                            extraParameters: [:]
                        )
                    )
                    switch onEnum(of: result) {
                    case .success:
                        await clearCredentials()
                        loginMethod = nil
                        state = .idle
                    // Logout did not complete: leave the user on the Welcome screen
                    // since the session was not cleared.
                    case .failure:
                        break
                    }
                } catch {
                    // Thrown error (e.g. cancellation): leave the view unchanged.
                }
            }
        } else {
            // Embedded login holds no browser session, so there is no round-trip:
            // just drop the stored credentials and return to the chooser.
            Task {
                await clearCredentials()
                loginMethod = nil
                state = .idle
            }
        }
    }

    // Persists credentials to the Keychain so a later launch can restore the session.
    // saveCredentials is `async throws` on the Swift side (SKIE bridges the Kotlin
    // suspend fun); a storage failure comes back as Result.Failure. This is a sample,
    // so a save failure is not surfaced: the user stays logged in for this run and
    // would only need to sign in again on the next launch.
    private func saveCredentials(_ credentials: Credentials) async {
        guard let credentialsManager else { return }
        do {
            _ = try await credentialsManager.saveCredentials(credentials: credentials)
        } catch {
            // Cancellation or an unexpected throw: ignore, the session stays valid in-memory.
        }
    }

    // Removes the stored credentials on logout so the next launch starts at the chooser.
    private func clearCredentials() async {
        guard let credentialsManager else { return }
        do {
            _ = try await credentialsManager.clearCredentials()
        } catch {
            // Nothing stored / cancellation: ignore.
        }
    }

    deinit {
        client?.close()
        webClient?.close()
    }
}

File: sample-app/androidApp/src/main/java/com/auth0/kmp/sample/AuthViewModel.kt
package com.auth0.kmp.sample

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.auth0.kmp.authentication.authenticationClient
import com.auth0.kmp.authentication.model.AuthParamsPublicKey
import com.auth0.kmp.authentication.model.AuthnParamsPublicKey
import com.auth0.kmp.authentication.model.DatabaseUser
import com.auth0.kmp.authentication.model.PublicKeyCredentials
import com.auth0.kmp.authentication.model.SignupProfile
import com.auth0.kmp.core.Auth0Account
import com.auth0.kmp.core.NetworkLogLevel
import com.auth0.kmp.core.NetworkingConfiguration
import com.auth0.kmp.core.RequestOptions
import com.auth0.kmp.core.error.Auth0Error
import com.auth0.kmp.core.model.Credentials
import com.auth0.kmp.core.result.Result
import com.auth0.kmp.credentials.credentialsManager
import com.auth0.kmp.webauth.LoginOptions
import com.auth0.kmp.webauth.LogoutOptions
import com.auth0.kmp.webauth.webAuthClient
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

sealed interface LoginUiState {
    // Initial state while the app checks storage for a saved session; the Splash
    // screen is shown until this resolves to Success (restored) or Idle (none).
    data object Restoring : LoginUiState
    data object Idle : LoginUiState
    data object Loading : LoginUiState
    data class Success(val credentials: Credentials) : LoginUiState
    data class Failure(val error: Auth0Error) : LoginUiState
}

// Drives the sign-up screen only. Kept separate from LoginUiState so a successful
// createUser (which does NOT mint tokens) doesn't trip the global Success->Welcome
// navigation; the confirmation screen shows the DatabaseUser, then the user opts
// in to logging in.
sealed interface SignupUiState {
    data object Idle : SignupUiState
    data object Loading : SignupUiState
    data class Success(val user: DatabaseUser) : SignupUiState
    data class Failure(val error: Auth0Error) : SignupUiState
}

private enum class LoginMethod { Embedded, WebAuth, Passkey }

// Wraps a non-Auth0 failure (a cancelled/failed passkey ceremony) as an Auth0Error
// so the shared Failure states can render it via toString(), like SDK errors.
private data class CeremonyError(val message: String) : Auth0Error {
    override fun toString(): String = message
}

class AuthViewModel(domain: String, clientId: String) : ViewModel() {

    // Whether the tenant config was supplied. When false we never build the SDK
    // client, so the app can still render its screens instead of crashing on a
    // blank domain (the SDK validates and throws eagerly, by design).
    val isConfigured: Boolean = domain.isNotBlank() && clientId.isNotBlank()

    private val account = if (isConfigured) {
        Auth0Account(
            clientId = clientId,
            domain = domain,
            configuration = NetworkingConfiguration(logLevel = NetworkLogLevel.BODY),
        )
    } else {
        null
    }

    private val audience = "https://firstresourceserver/"
    private val client = account?.let { authenticationClient(it) }
    private val webClient = account?.let { webAuthClient(it) }
    private val credentialsManager = account?.let { credentialsManager(it) }

    private val _state = MutableStateFlow<LoginUiState>(LoginUiState.Restoring)
    val state: StateFlow<LoginUiState> = _state.asStateFlow()

    private val _signupState = MutableStateFlow<SignupUiState>(SignupUiState.Idle)
    val signupState: StateFlow<SignupUiState> = _signupState.asStateFlow()

    // Retains the sign-up inputs so the confirmation screen's "Log in" can auto-
    // authenticate the just-created user without re-prompting for a password.
    private var lastSignup: Triple<String, String, String>? = null

    // Tracks how the current session was established, so logout only performs the
    // browser round-trip for Web Auth sessions (embedded login holds no SSO cookie).
    private var loginMethod: LoginMethod? = null

    // Checks storage for a saved session on launch. The chooser is shown right
    // away so cold start never blocks on a token round-trip; the restore then
    // runs in the background. A fresh token promotes to the logged-in screen via
    // a local decrypt (no network); an expired-but-refreshable token is renewed
    // off the launch path and then promotes. Anything unrestorable (no
    // credentials, or an expired token with no refresh token) simply stays on
    // the chooser — that is not an error to surface, the user just signs in.
    fun restoreSession() {
        val credentialsManager = credentialsManager
        if (credentialsManager == null) {
            _state.value = LoginUiState.Idle
            return
        }
        _state.value = LoginUiState.Idle
        viewModelScope.launch {
            val result = credentialsManager.getCredentials()
            if (result is Result.Success) {
                _state.value = LoginUiState.Success(result.data)
            }
        }
    }

    fun login(
        email: String,
        password: String,
        realm: String,
        options: RequestOptions = RequestOptions(),
    ) {
        val client = client ?: return
        _state.value = LoginUiState.Loading
        viewModelScope.launch {
            // Request offline_access so the tenant issues a refresh token, letting
            // getCredentials() renew an expired access token on a later launch.
            val result = client.login(
                usernameOrEmail = email,
                password = password,
                realm = realm,
                audience = audience,
                scope = "openid profile email offline_access",
                options = options,
            )
            _state.value = when (result) {
                is Result.Success -> {
                    loginMethod = LoginMethod.Embedded
                    credentialsManager?.saveCredentials(result.data)
                    LoginUiState.Success(result.data)
                }

                is Result.Failure -> LoginUiState.Failure(result.error)
            }
        }
    }

    // Creates a database user, then parks the result on the confirmation screen.
    // No tokens are issued here (createUser returns a DatabaseUser, not
    // Credentials), so the session state is untouched; the user chooses to log in
    // next via completeSignupLogin().
    fun createUser(
        email: String,
        password: String,
        connection: String,
        options: RequestOptions = RequestOptions(),
    ) {
        val client = client ?: return
        _signupState.value = SignupUiState.Loading
        viewModelScope.launch {
            val result = client.createUser(
                profile = SignupProfile(email = email),
                password = password,
                connection = connection,
                options = options,
            )
            _signupState.value = when (result) {
                is Result.Success -> {
                    lastSignup = Triple(email, password, connection)
                    SignupUiState.Success(result.data)
                }

                is Result.Failure -> SignupUiState.Failure(result.error)
            }
        }
    }

    // Logs in the user that was just created, reusing the retained sign-up inputs.
    // Drives the shared LoginUiState so success navigates to Welcome like any login.
    fun completeSignupLogin() {
        val (email, password, connection) = lastSignup ?: return
        login(email = email, password = password, realm = connection)
    }

    fun resetSignup() {
        _signupState.value = SignupUiState.Idle
        lastSignup = null
    }

    // Registers a passkey, then mints tokens by sending the registration credential
    // and the signup's auth_session straight to the passkey token grant — no
    // separate login challenge (mirrors Auth0.Android's signup->signinWithPasskey).
    // runCeremony is supplied by the screen because the WebAuthn UI must be anchored
    // to an Activity.
    fun passkeySignup(
        email: String,
        connection: String,
        runCeremony: suspend (AuthnParamsPublicKey) -> PublicKeyCredentials,
    ) {
        val client = client ?: return
        _state.value = LoginUiState.Loading
        viewModelScope.launch {
            val challenge = client.passkeySignupChallenge(
                profile = SignupProfile(email = email),
                realm = connection,
            )
            when (challenge) {
                is Result.Failure -> {
                    _state.value = LoginUiState.Failure(challenge.error)
                    return@launch
                }

                is Result.Success -> {
                    val credential = try {
                        runCeremony(challenge.data.authParamsPublicKey)
                    } catch (e: Exception) {
                        _state.value = LoginUiState.Failure(
                            CeremonyError("Passkey registration was cancelled or failed: ${e.message}"),
                        )
                        return@launch
                    }
                    val result = client.loginWithPasskey(
                        authSession = challenge.data.authSession,
                        authResponse = credential,
                        realm = connection,
                    )
                    _state.value = when (result) {
                        is Result.Success -> {
                            loginMethod = LoginMethod.Passkey
                            credentialsManager?.saveCredentials(result.data)
                            LoginUiState.Success(result.data)
                        }

                        is Result.Failure -> LoginUiState.Failure(result.error)
                    }
                }
            }
        }
    }

    // Requests a login challenge, runs the assertion ceremony, then exchanges the
    // result for tokens. runCeremony is supplied by the screen (Activity-anchored).
    fun passkeyLogin(
        connection: String,
        runCeremony: suspend (AuthParamsPublicKey) -> PublicKeyCredentials,
    ) {
        val client = client ?: return
        _state.value = LoginUiState.Loading
        viewModelScope.launch {
            val challenge = client.passkeyLoginChallenge(realm = connection)
            when (challenge) {
                is Result.Failure -> {
                    _state.value = LoginUiState.Failure(challenge.error)
                    return@launch
                }

                is Result.Success -> {
                    val credential = try {
                        runCeremony(challenge.data.authParamsPublicKey)
                    } catch (e: Exception) {
                        _state.value = LoginUiState.Failure(
                            CeremonyError("Passkey sign-in was cancelled or failed: ${e.message}"),
                        )
                        return@launch
                    }
                    val result = client.loginWithPasskey(
                        authSession = challenge.data.authSession,
                        authResponse = credential,
                        realm = connection,
                    )
                    _state.value = when (result) {
                        is Result.Success -> {
                            loginMethod = LoginMethod.Passkey
                            credentialsManager?.saveCredentials(result.data)
                            LoginUiState.Success(result.data)
                        }

                        is Result.Failure -> LoginUiState.Failure(result.error)
                    }
                }
            }
        }
    }

    fun webLogin() {
        val webClient = webClient ?: return
        _state.value = LoginUiState.Loading
        viewModelScope.launch {
            _state.value = when (val result = webClient.login(LoginOptions(audience = audience))) {
                is Result.Success -> {
                    loginMethod = LoginMethod.WebAuth
                    credentialsManager?.saveCredentials(result.data)
                    LoginUiState.Success(result.data)
                }

                is Result.Failure -> LoginUiState.Failure(result.error)
            }
        }
    }

    fun logout() {
        val webClient = webClient
        if (loginMethod == LoginMethod.WebAuth && webClient != null) {
            // Keep the current Success state (credentials stay on screen) during the
            // browser round-trip; only clear the view once logout actually succeeds.
            viewModelScope.launch {
                when (webClient.logout(LogoutOptions())) {
                    is Result.Success -> {
                        credentialsManager?.clearCredentials()
                        loginMethod = null
                        _state.value = LoginUiState.Idle
                    }
                    // Logout did not complete (cancelled/failed): leave the user on
                    // the Welcome screen since the session was not cleared.
                    is Result.Failure -> Unit
                }
            }
        } else {
            // Embedded (or restored) session: no SSO cookie to clear in the browser,
            // so just drop the stored credentials locally.
            loginMethod = null
            viewModelScope.launch {
                credentialsManager?.clearCredentials()
                _state.value = LoginUiState.Idle
            }
        }
    }

    override fun onCleared() {
        super.onCleared()
        client?.close()
        webClient?.close()
    }
}
